Blog: Locality failover for external services (#9880)
istio/istio.io: content/en/blog/2021/external-locality-failover/index.md

Similar to services running inside the service mesh, you can configure Istio to detect outliers and failover to a healthy endpoint, while still being completely transparent to your application. In this example, we’ll use Amazon DynamoDB endpoints and pick a primary region that is the same or close to workloads running in a Google Kubernetes Engine (GKE) cluster. We’ll also configure a failover region.

|Routing|Endpoint|
|--- |--- |
|Primary|http://dynamodb.us-east-1.amazonaws.com|
|Failover|http://dynamodb.us-west-1.amazonaws.com|

![failover](./external-locality-failover.png)

## Define external endpoints using a ServiceEntry

First, determine the location of your workloads.

{{< text bash >}}
$ kubectl describe node | grep failure-domain.beta.kubernetes.io/region
                    failure-domain.beta.kubernetes.io/region=us-east1
                    failure-domain.beta.kubernetes.io/region=us-east1
{{< /text >}}

> In this example, the GKE cluster nodes are running in `us-east1`

Reviewer: ```suggestion
In this example, the GKE cluster nodes are running in `us-east1`.
```

> First, determine the location of your workloads.
> 
> {{< text bash >}}
> $ kubectl describe node | grep failure-domain.beta.kubernetes.io/region
>                     failure-domain.beta.kubernetes.io/region=us-east1
>                     failure-domain.beta.kubernetes.io/region=us-east1
> {{< /text >}}
> 
> In this example, the GKE cluster nodes are running in `us-east1`.

Reviewer: ```suggestion
[Locality load balancing](/docs/tasks/traffic-management/locality-load-balancing/) works based on `region` or `zone`, which are usually inferred from labels set on the Kubernetes nodes.  First, determine the location of your workloads:

{{< text bash >}}
$ kubectl describe node | grep topology.kubernetes.io/region
                    topology.kubernetes.io/region=us-east1
{{< /text >}}

In this example, the GKE cluster nodes are running in the `us-east1` region.
```

> Next, apply a `ServiceEntry` to create a `mydb.com` service that’s backed by the two DynamoDB endpoints. Set the `locality` of your primary endpoint to the same region as your workload.

Reviewer: ```suggestion
Next, create a `ServiceEntry` for an arbitrary `mydb.com` service, backed by the two DynamoDB endpoints. Set the `locality` of the primary endpoint to the same region as your workload:
```

{{< text yaml >}}
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
 name: external-svc-dns
spec:
 hosts:
 - mydb.com
 location: MESH_EXTERNAL
 ports:
 - number: 80
   name: http
   protocol: HTTP
 resolution: DNS
 endpoints:
 - address: dynamodb.us-east-1.amazonaws.com
   locality: us-east1
   ports:
     http: 80
 - address: dynamodb.us-west-1.amazonaws.com

>    locality: us-west

Reviewer: one more nit: one is us-east*1*, the other is us-west, throughout?

(This might be confusing Google regions with AWS regions)

Author: because the cluster nodes are only in `us-east1`, it doesn't matter what you set the locality of the other endpoints to. I didn't set it to `us-west1` because that seemed oddly specific as it doesn't map to any value.

Reviewer: `other`? `failover`?

---
title: "Configuring failover for external services"
description: Learn how to configure locality load balancing and failover for endpoints that are outside of your mesh.
publishdate: 2021-06-04
attribution: "Ram Vennam (Solo.io)"
keywords: [locality,region,failover,Istio,outlier,external]
---

> Istio’s powerful APIs can be used to solve a variety of service mesh use cases. Many users know about it’s strong ingress and east-west capabilities but it also offers many features for egress (outgoing) traffic. This is especially useful when your application needs to talk to an external service - such as a database endpoint provided by a cloud provider. There are often choices of [multiple endpoints](https://docs.aws.amazon.com/general/latest/gr/ddb.html) depending on where your workload is running. You typically want to choose the endpoint closest to your workload for latency reasons, but you may need to configure automatic failover to another endpoint in case things are not working as expected.

Reviewer: > about it’s

its

> There are often choices of [multiple endpoints](https://docs.aws.amazon.com/general/latest/gr/ddb.html)

Only on clouds without good anycast load balancers :)
That's not a general link for "multiple endpoints", so it might pay to introduce the fact you're talking about DynamoDB earlier.

"For example, Amazon's DynamoDB offers different endpoints depending on where your workload is running".

(I assume it's different endpoints to the same global service, not regional slices of it?)

> in case

in _the_ case?  (unsure)

Similar to services running inside the service mesh, you can configure Istio to detect outliers and failover to a healthy endpoint, while still being completely transparent to your application. In this example, we’ll use Amazon DynamoDB endpoints and pick a primary region that is the same or close to workloads running in a Google Kubernetes Engine (GKE) cluster. We’ll also configure a failover region.

|Routing|Endpoint|
|--- |--- |
|Primary|http://dynamodb.us-east-1.amazonaws.com|
|Failover|http://dynamodb.us-west-1.amazonaws.com|

![failover](./external-locality-failover.png)

## Define external endpoints using a ServiceEntry

First, determine the location of your workloads.

{{< text bash >}}
$ kubectl describe node | grep failure-domain.beta.kubernetes.io/region
                    failure-domain.beta.kubernetes.io/region=us-east1
                    failure-domain.beta.kubernetes.io/region=us-east1
{{< /text >}}

> As you can see, on my GKE cluster, all of my nodes are running in `us-east1`

Reviewer: The blog refers to `you` and `your` workloads, so write "In this example", rather than using `me`?

> Next, apply a Istio ServiceEntry to create a `mydb.com` service that’s backed by the two DynamoDB endpoints.

Reviewer: `ServiceEntry`

{{< text yaml >}}

> apiVersion: networking.istio.io/v1alpha3

Reviewer: v1beta1

kind: ServiceEntry
metadata:
 name: external-svc-dns
spec:
 hosts:
 - mydb.com
 location: MESH_EXTERNAL
 ports:
 - number: 80
   name: http
   protocol: HTTP
 resolution: DNS
 endpoints:
 - address: dynamodb.us-east-1.amazonaws.com
   locality: us-east1
   ports:
     http: 80
 - address: dynamodb.us-west-1.amazonaws.com
   locality: us-west1
   ports:
     http: 80
{{< /text >}}

> Let’s deploy a sleep container to runs some tests:

Reviewer: Personally I like to explain why we use a sleep container, or don't even mention it.  "Deploy a debugging container" or something

{{< text bash >}}
$ kubectl apply -f @samples/sleep/sleep.yaml@
{{< /text >}}

From the sleep container try going to `http://mydb.com` 5 times:

{{< text bash >}}
$ for i in {1..5}; do kubectl exec deploy/sleep -c sleep -- curl -sS http://mydb.com; echo; sleep 2; done
healthy: dynamodb.us-east-1.amazonaws.com
healthy: dynamodb.us-west-1.amazonaws.com
healthy: dynamodb.us-west-1.amazonaws.com
healthy: dynamodb.us-east-1.amazonaws.com
healthy: dynamodb.us-east-1.amazonaws.com
{{< /text >}}

> It’s sending requests to both endpoints. We only want it to send to the endpoint marked with the same region as our nodes.

Reviewer: "You will see that Istio is sending" ?

For that, we need to configure a `DestinationRule`.

## Set failover conditions using a `DestinationRule`

Istio’s `DestinationRule` lets you configure load balancing, connection pool, and outlier detection settings. We can specify the conditions used to identify an endpoint as unhealthy and remove it from the load balancing pool.

{{< text yaml >}}
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
 name: mydynamodb
spec:
 host: mydb.com
 trafficPolicy:
   outlierDetection:
     consecutive5xxErrors: 1
     interval: 15s
     baseEjectionTime: 1m
{{< /text >}}

> The above `DestinationRule` configures the endpoints to be scanned every 15 seconds, and if any endpoint fails with a 5xx error code even once, it will be marked unhealthy for one minute. If this circuit breaker is not triggered, the traffic will route to the same region as the pod.

Reviewer: "This rule configures"

code, even once

If we run our curl again, we should see that traffic is always going to the `us-east1` endpoint.

{{< text bash >}}
$ for i in {1..5}; do kubectl exec deploy/sleep -c sleep -- curl -sS http://mydb.com; echo; sleep 2; done

healthy: dynamodb.us-east-1.amazonaws.com
healthy: dynamodb.us-east-1.amazonaws.com
healthy: dynamodb.us-east-1.amazonaws.com
healthy: dynamodb.us-east-1.amazonaws.com
healthy: dynamodb.us-east-1.amazonaws.com
{{< /text >}}

## Simulate a failure

> Next, let's see what happens if the us-east endpoint goes down. To simulate this, let’s modify the ServiceEntry and set the `us-east` endpoint to an invalid port (change port from 80 to 81).

Reviewer: > port (change port from 80 to 81).

I think just saying "an invalid port:" or "an erroneous port:" would be ok

{{< text yaml >}}
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
 name: mydynamodb
spec:
 host: mydb.com
 trafficPolicy:
   tls:
     mode: SIMPLE
   loadBalancer:
     simple: ROUND_ROBIN
     localityLbSetting:
       enabled: true
       failover:
         - from: us-east1
           to: us-west
   outlierDetection:
     consecutive5xxErrors: 1
     interval: 15s
     baseEjectionTime: 1m
{{< /text >}}

 The `DestinationRule` now performs TLS origination and configures the outlier detection. The rule also has a [failover](/docs/reference/config/networking/destination-rule/#LocalityLoadBalancerSetting) field configured where you can specify exactly what regions are failover targets. This is useful when you have several regions defined.

## **Wrapping Up**

Istio’s `VirtualService` and `DestinationRule` API’s provide traffic routing, failure recovery and fault injection features so that you can create resilient applications. The ServiceEntry API extends many of these features to external services that are not part of your service mesh.

> As a Field Engineer at [Solo.io](https://www.solo.io/), I help users solve interesting problems such as these using Istio. If you have any questions, feel free to reach out to me on our [Slack](https://slack.solo.io/).

Reviewer: I think you could make this point more organically throughout the post :)

> As a Field Engineer at [Solo.io](https://www.solo.io/), I help users solve interesting problems such as these using Istio. If you have any questions, feel free to reach out to me on our [Slack](https://slack.solo.io/).

Reviewer 2: I think we should discourage these CTAs on istio.io blog..
